### PR TITLE
Skill system revamp

### PR DESCRIPTION
**Pull request by Ilysen**
#### purpose

skills are super old right now, and unfortunately it shows. even if they're left as completely aesthetic like they are currently, the system's been long in need of some care, and this is a first step towards hopefully makin' that happen

#### details

**NOTE FOR NON-CODERS:** this revamp, as of right now, is still purely aesthetic. though this might change before the final iteration, the intent is to make more modern code and skill definitions while not affecting gameplay, at least for now.

anyway, the nitty gritty:
* introduces a new type of datum for tracking skills and creates a bunch of skills using that datum. currently this is called `/datum/new_skill`, but this is placeholder and will replace the old skill datum once it's time to rip them out
* instead of being measured on a scale from Untrained to Professional, most skills just represent proficiency with their specific thing. having the `Atmospherics` skill, for instance, implies that you have good knowledge of pipes and so on, while not having it implies that you don't.
  * there are some exceptions to this - some skills, like Anatomy and Close Combat, have multiple ranks to represent varying degrees of competence. someone with Anatomy 1 could be thought to be a novice at surgery without much practical experience, while someone with Anatomy 2 is much more comfortable performing operations.
* skills are tied to mind, rather than mob
* each job can come with its own set of skills that will be given to the player for free upon joining as that job

job code is pretty butts right now so in a way this is putting the cart before the horse, but I only realized this when I was midway through making the foundational code for this PR and I'm not in the habit of throwing stuff out unless I really need to :p

#### testing

WIP. will need to be extensive

#### todo

- [x] foundational work - make sure it compiles and the datums are there
- [x] skill distributions for all standard jobs
- [ ] skill menu to see what you have mid-round
- [ ] ability to pick out skills in chargen like you currently can, ideally taking job skills into account as well
  - [ ] if possible, allow skills to be customized per-job - that way people can play multiple jobs on one character with different skillsets for each. will need to learn the save/load system for this
- [ ] polish pass/skill audit; pruning, expanding as needed, and so on
- [ ] stretch goal: skills for alt titles. I had trouble getting this working, but it's a desired thing so that jobs like Virologist specifically have the Virology skill without needing to pick it up manually, but as it's all aesthetic right now it's not too big of a deal I think

### CODE REVIEW

**Review by Atermonera — COMMENTED** [review]
> ability to pick out skills in chargen like you currently can, ideally taking job skills into account as well

This I think will be really clunky unless we rewrite character setup to have per-job configuration (Like loadouts)

> having the `Atmospherics` skill, for instance, implies that you have good knowledge of pipes and so on, while not having it implies that you don't.

I think a binary option makes sense in several cases, but being binary in this manner makes sequential tiers (Your example of cooking) kinda... weird. Could we make it so that each skill can define its own levels, such that you can make some skills be binary and others have proper tiers?

Having read most of the skill descriptions, I'm pretty dubious of a number of dependencies, descriptions, and potential mechanical interactions. Not commenting on them here because this first review is looking at the foundation of the system overall.

File: code/datums/mind.dm [hunk 1]
```diff
@@ -497,6 +500,42 @@
 	if(G)
 		G.reenter_corpse()
 
+/// Checks if this mind has a skill tree's parent node unlocked.
+/datum/mind/proc/has_skill_tree(tree_id)
+	for (var/datum/new_skill/S in skills)
+		if (S.tree == tree_id && S.tree_parent)
+			return TRUE
```
[the remark this answers is not in the document]
Comment (Ilysen) [reply code/datums/mind.dm:507]: it does! I opted for it to be checked anyway in the case of a tree being made whose name does not match its ID, but if you like, I can do that.

File: code/datums/mind.dm [hunk 2]
```diff
@@ -497,6 +500,42 @@
...
+
+/// Adds a skill to this mind if it doesn't have one.
+/// If `rank` is defined, will also upgrade existing skills to the provided rank.
+/datum/mind/proc/add_skill(skill_id, rank = 1)
+	if (has_skill(skill_id))
+		skills[skill_id].rank = max(skills[skill_id].rank, rank)
+		return
+	for (var/ST in subtypesof(/datum/new_skill))
+		var/datum/new_skill/S = ST
+		if (initial(S.id) == skill_id)
```
[the remark this answers is not in the document]
Comment (Ilysen) [reply code/datums/mind.dm:522]: perhaps. what would you have in mind - something like an associative list containing all skill typepaths that can be quickly referenced using the string ID?

**Comment by Ilysen** [comment]
> This I think will be really clunky unless we rewrite character setup to have per-job configuration (Like loadouts)

I can do that if need be :p

> I think a binary option makes sense in several cases, but being binary in this manner makes sequential tiers (Your example of cooking) kinda... weird. Could we make it so that each skill can define its own levels, such that you can make some skills be binary and others have proper tiers?

I'm not certain if you saw, but there's support for this. skills can have arbitrary additional ranks defined, and some skills that do that right now include anatomy, piloting, and the weapon skills. heck, I could make every skill have a rank if so desired

> Having read most of the skill descriptions, I'm pretty dubious of a number of dependencies, descriptions, and potential mechanical interactions. Not commenting on them here because this first review is looking at the foundation of the system overall.

I'm happy to change any of the particular descriptions; right now, it's meant to be only aesthetic, like the existing skill system (a notable exception here are the "proficiency" skills, which I currently think I'll be cutting out.) at the moment you can perform any task in-game even if you're unskilled in the relevant skill, so this doesn't have any changes to that - it just makes it cleaner.

File: code/datums/mind.dm (hunk 2 above)
[the remark this answers is not in the document]
Comment (Atermonera) [reply code/datums/mind.dm:524]: I'm proposing dropping the string ID altogether and using the typepath directly in its place.

File: code/datums/mind.dm (hunk 1 above)
[the remark this answers is not in the document]
Comment (Atermonera) [reply code/datums/mind.dm:507]: Per my other comment, if the "ID" is the typepath, then there is exactly 1 type that can correctly resolve to that path.

File: code/datums/mind.dm (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ilysen) [reply code/datums/mind.dm:524]: I can do that. I'd worried it might be kind of unclean, but if you're cool with it, then I can go that route.

**Comment by Ilysen** [comment]
swapped the string ID logic for typepath logic. I'm not nearly as experienced with typepaths, so holler at me if there's any immediate issues. have not had time to test beyond very briefly booting up the example map on a character with the CMO job and calling procs in the mind datum via VV, but those two things seemed to work :p

**Comment by Ilysen** [comment]
ah hecc, I've been chipping away at trying to get a UI set up for this and my skills just aren't up to snuff with html. the added complexity of this system - trees and ranks and whatnot - is really difficult to represent with what I know (since I don't know if it's feasible to make use of things like dropdowns, tooltips, and so on) and I've gradually become disillusioned with the rank system anyway. I think the skill system needs a revamp, but I think this wasn't how I shoulda gone about it. sorry for wasting your time :<

**Pull request closed without merging**